File: mobileapp/lib/services/app_api_service.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:omphysioworld/config/api_config.dart';
import 'package:omphysioworld/storage/patient_session_storage.dart';

class ApiException implements Exception {
  ApiException(this.message);

  final String message;

  @override
  String toString() => message;
}

class AppUploadFile {
  const AppUploadFile({
    required this.name,
    required this.bytes,
    this.mimeType,
  });

  final String name;
  final Uint8List bytes;
  final String? mimeType;
}

class AppApiService {
  AppApiService({
    HttpClient? client,
    PatientSessionStorage? patientSessionStorage,
  })  : _client = client ?? HttpClient(),
        _patientSessionStorage = patientSessionStorage ?? PatientSessionStorage();

  final HttpClient _client;
  final PatientSessionStorage _patientSessionStorage;

  Future<Map<String, dynamic>> healthCheck() {
    return _getJson('/health');
  }

  Future<Map<String, dynamic>> login({
    required String email,
    required String password,
  }) {
    return _postJson(
      '/auth/login',
      {
        'email': email,
        'password': password,
      },
    );
  }

  Future<Map<String, dynamic>> register({
    required String name,
    required String email,
    required String mobile,
    required String password,
  }) {
    return _postJson(
      '/auth/register',
      {
        'name': name,
        'email': email,
        'mobile': mobile,
        'password': password,
        'createdFrom': 'mobile_app',
      },
    );
  }

  Future<Map<String, dynamic>> requestPasswordReset({
    required String email,
  }) {
    return _postJson(
      '/auth/forgot-password',
      {
        'email': email,
      },
    );
  }

  Future<Map<String, dynamic>> submitAppointment({
    required String name,
    required String email,
    required String phone,
    String patientId = '',
    required String service,
    required String date,
    String time = '',
    required String message,
  }) {
    return _postJson(
      '/appointments',
      {
        'name': name,
        'email': email,
        'phone': phone,
        'patientId': patientId,
        'service': service,
        'date': date,
        'time': time,
        'message': message,
      },
    );
  }

  Future<Map<String, dynamic>> submitContactMessage({
    required String name,
    required String email,
    required String phone,
    required String subject,
    required String message,
  }) {
    return _postJson(
      '/contact',
      {
        'name': name,
        'email': email,
        'phone': phone,
        'subject': subject,
        'message': message,
      },
    );
  }

  Future<Map<String, dynamic>> submitStaffApplication({
    required String name,
    required String email,
    required String phone,
    required String role,
    required String experience,
    required String message,
    AppUploadFile? resume,
  }) {
    return _postMultipart(
      '/staff-applications',
      fields: {
        'name': name,
        'email': email,
        'phone': phone,
        'role': role,
        'experience': experience,
        'message': message,
      },
      files: resume == null ? const [] : [resume],
      fileFieldName: 'resume',
    );
  }

  Future<Map<String, dynamic>> submitPatientClinicalNote({
    required String patientId,
    required String title,
    required String note,
    String addedByLabel = 'Patient',
    List<AppUploadFile> documents = const [],
  }) {
    return _postMultipart(
      '/patients/$patientId/clinical-notes',
      fields: {
        'title': title,
        'note': note,
        'addedByType': 'patient',
        'addedByLabel': addedByLabel,
      },
      files: documents,
    );
  }

  Future<Map<String, dynamic>> uploadPatientProfileImage({
    required String patientId,
    required AppUploadFile image,
  }) {
    return _postMultipart(
      '/patients/$patientId/profile-image',
      fields: const {},
      files: [image],
      fileFieldName: 'image',
    );
  }

  Future<Map<String, dynamic>> getPatient({
    required String patientId,
  }) {
    return _getJson('/patients/$patientId');
  }

  Future<Map<String, dynamic>> updatePatientProfile({
    required String patientId,
    required String name,
    required String mobile,
    required String disease,
  }) {
    return _putJson(
      '/patients/$patientId',
      {
        'name': name,
        'mobile': mobile,
        'disease': disease,
      },
    );
  }

  Future<Map<String, dynamic>> changePassword({
    required String oldPassword,
    required String newPassword,
    required String confirmPassword,
  }) {
    return _postJson(
      '/auth/change-password',
      {
        'oldPassword': oldPassword,
        'newPassword': newPassword,
        'confirmPassword': confirmPassword,
      },
    );
  }

  Future<List<Map<String, dynamic>>> getServices() async {
    final response = await _getJson('/services');
    final data = response['data'];

    if (data is List) {
      return data.whereType<Map<String, dynamic>>().toList();
    }

    return const [];
  }

  Future<List<Map<String, dynamic>>> getPatientAppointmentRequests({
    required String patientId,
  }) async {
    final response = await _getJson('/patients/$patientId/appointment-requests');
    final data = response['data'];

    if (data is List) {
      return data.whereType<Map<String, dynamic>>().toList();
    }

    return const [];
  }

  Future<String> resolveResourceUrl(String pathOrUrl) async {
    final value = pathOrUrl.trim();
    if (value.isEmpty || value.startsWith('http://') || value.startsWith('https://')) {
      return value;
    }

    final baseUrl = _resolveBaseUrl();
    return value.startsWith('/') ? '$baseUrl$value' : '$baseUrl/$value';
  }

  Future<List<Map<String, dynamic>>> getPublicChatAgents() async {
    final response = await _getJson('/public-chat-agents');
    final data = response['data'];

    if (data is List) {
      return data.whereType<Map<String, dynamic>>().toList();
    }

    return const [];
  }

  Future<Map<String, dynamic>> getPublicChatConversation({
    required String conversationId,
  }) {
    return _getJson('/public-chat/conversations/$conversationId');
  }

  Future<Map<String, dynamic>> startPublicChatConversation({
    required String agentId,
    required String visitorName,
    required String visitorContact,
    required String text,
    List<AppUploadFile> attachments = const [],
  }) {
    return _postMultipart(
      '/public-chat/conversations',
      fields: {
        'agentId': agentId,
        'visitorName': visitorName,
        'visitorContact': visitorContact,
        'text': text,
      },
      files: attachments,
      fileFieldName: 'attachments',
    );
  }

  Future<Map<String, dynamic>> sendPublicChatMessage({
    required String conversationId,
    required String visitorName,
    required String text,
    List<AppUploadFile> attachments = const [],
  }) {
    return _postMultipart(
      '/public-chat/conversations/$conversationId/messages',
      fields: {
        'visitorName': visitorName,
        'text': text,
      },
      files: attachments,
      fileFieldName: 'attachments',
    );
  }

  Future<Map<String, dynamic>> _postJson(
    String path,
    Map<String, dynamic> payload,
  ) async {
    return _guardRequest(() async {
      final baseUrl = _resolveBaseUrl();
      final request = await _client.postUrl(Uri.parse('$baseUrl$path'));
      request.headers.contentType = ContentType.json;
      await _applyAuthHeader(request);
      request.write(jsonEncode(payload));

      return _readJsonResponse(response: await request.close());
    });
  }

  Future<Map<String, dynamic>> _getJson(
    String path,
  ) async {
    return _guardRequest(() async {
      final baseUrl = _resolveBaseUrl();
      final request = await _client.getUrl(Uri.parse('$baseUrl$path'));
      await _applyAuthHeader(request);
      return _readJsonResponse(response: await request.close());
    });
  }

  Future<Map<String, dynamic>> _putJson(
    String path,
    Map<String, dynamic> payload,
  ) async {
    return _guardRequest(() async {
      final baseUrl = _resolveBaseUrl();
      final request = await _client.putUrl(Uri.parse('$baseUrl$path'));
      request.headers.contentType = ContentType.json;
      await _applyAuthHeader(request);
      request.write(jsonEncode(payload));

      return _readJsonResponse(response: await request.close());
    });
  }

  Future<Map<String, dynamic>> _postMultipart(
    String path, {
    required Map<String, String> fields,
    required List<AppUploadFile> files,
    String fileFieldName = 'documents',
  }) async {
    return _guardRequest(() async {
      final baseUrl = _resolveBaseUrl();
      final boundary = '----ommphysio-${DateTime.now().microsecondsSinceEpoch}';
      final request = await _client.postUrl(Uri.parse('$baseUrl$path'));
      request.headers.contentType = ContentType(
        'multipart',
        'form-data',
        parameters: {'boundary': boundary},
      );
      await _applyAuthHeader(request);

      for (final entry in fields.entries) {
        _writeMultipartTextField(
          request: request,
          boundary: boundary,
          name: entry.key,
          value: entry.value,
        );
      }

      for (final file in files) {
        _writeMultipartFile(
          request: request,
          boundary: boundary,
          fieldName: fileFieldName,
          file: file,
        );
      }

      request.add(utf8.encode('--$boundary--\r\n'));
      return _readJsonResponse(response: await request.close());
    });
  }

  void _writeMultipartTextField({
    required HttpClientRequest request,
    required String boundary,
    required String name,
    required String value,
  }) {
    request.add(
      utf8.encode(
        '--$boundary\r\n'
        'Content-Disposition: form-data; name="${_escapeMultipartValue(name)}"\r\n'
        '\r\n'
        '$value\r\n',
      ),
    );
  }

  void _writeMultipartFile({
    required HttpClientRequest request,
    required String boundary,
    required String fieldName,
    required AppUploadFile file,
  }) {
    request.add(
      utf8.encode(
        '--$boundary\r\n'
        'Content-Disposition: form-data; name="${_escapeMultipartValue(fieldName)}"; filename="${_escapeMultipartValue(file.name)}"\r\n'
        'Content-Type: ${file.mimeType ?? _guessMimeType(file.name)}\r\n'
        '\r\n',
      ),
    );
    request.add(file.bytes);
    request.add(utf8.encode('\r\n'));
  }

  Future<Map<String, dynamic>> _readJsonResponse({
    required HttpClientResponse response,
  }) async {
    final responseBody = await utf8.decoder.bind(response).join();
    final decodedBody = _decodeJsonObject(responseBody);

    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw ApiException(
        decodedBody['message']?.toString() ??
            'Request failed with status ${response.statusCode}.',
      );
    }

    return decodedBody;
  }

  Future<T> _guardRequest<T>(Future<T> Function() request) async {
    try {
      return await request().timeout(const Duration(seconds: 25));
    } on ApiException {
      rethrow;
    } on SocketException {
      throw ApiException(
        'No internet connection. Please check your network and try again.',
      );
    } on TimeoutException {
      throw ApiException(
        'OPW is taking too long to respond. Please try again in a moment.',
      );
    } on HandshakeException {
      throw ApiException(
        'Secure connection failed. Please try again shortly.',
      );
    } on HttpException {
      throw ApiException(
        'Unable to reach OPW right now. Please try again shortly.',
      );
    } on FormatException {
      throw ApiException(
        'Unexpected response from OPW. Please try again.',
      );
    }
  }

  String _resolveBaseUrl() {
    return ApiConfig.defaultBaseUrl;
  }

  Future<void> _applyAuthHeader(HttpClientRequest request) async {
    final patientUser = await _patientSessionStorage.getPatientUser();
    final token = patientUser?['token']?.toString().trim() ?? '';

    if (token.isNotEmpty) {
      request.headers.set(HttpHeaders.authorizationHeader, 'Bearer $token');
    }
  }

  Map<String, dynamic> _decodeJsonObject(String responseBody) {
    if (responseBody.trim().isEmpty) {
      return <String, dynamic>{};
    }

    final decoded = jsonDecode(responseBody);
    if (decoded is Map<String, dynamic>) {
      return decoded;
    }

    return <String, dynamic>{
      'data': decoded,
    };
  }

  String _escapeMultipartValue(String value) {
    return value.replaceAll('"', '\\"');
  }

  String _guessMimeType(String fileName) {
    final lowerName = fileName.toLowerCase();
    if (lowerName.endsWith('.pdf')) {
      return 'application/pdf';
    }
    if (lowerName.endsWith('.png')) {
      return 'image/png';
    }
    if (lowerName.endsWith('.jpg') || lowerName.endsWith('.jpeg')) {
      return 'image/jpeg';
    }
    if (lowerName.endsWith('.webp')) {
      return 'image/webp';
    }
    return 'application/octet-stream';
  }
}

File: userapp/app/src/main/java/com/ommphysioworld/userapp/data/AppApiService.kt
package com.ommphysioworld.userapp.data

import com.ommphysioworld.userapp.BuildConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedInputStream
import java.io.DataOutputStream
import java.io.IOException
import java.net.HttpURLConnection
import java.net.SocketTimeoutException
import java.net.URL
import java.nio.charset.StandardCharsets
import javax.net.ssl.HttpsURLConnection
import javax.net.ssl.SSLException

class ApiException(
    message: String,
    val statusCode: Int? = null,
) : Exception(message)

data class UploadFile(
    val name: String,
    val bytes: ByteArray,
    val mimeType: String? = null,
)

class AppApiService(
    private val storage: AppStorage,
) {
    suspend fun login(email: String, password: String): JsonMap {
        return postJson(
            "/auth/login",
            mapOf(
                "email" to email,
                "password" to password,
            ),
        )
    }

    suspend fun register(name: String, email: String, mobile: String, password: String): JsonMap {
        return postJson(
            "/auth/register",
            mapOf(
                "name" to name,
                "email" to email,
                "mobile" to mobile,
                "password" to password,
                "createdFrom" to "mobile_app",
            ),
        )
    }

    suspend fun requestPasswordReset(email: String): JsonMap {
        return postJson(
            "/auth/forgot-password",
            mapOf("email" to email),
        )
    }

    suspend fun submitAppointment(
        name: String,
        email: String,
        phone: String,
        patientId: String,
        service: String,
        date: String,
        time: String,
        message: String,
        documents: List<UploadFile>,
    ): JsonMap {
        return postMultipart(
            "/appointments",
            fields = mapOf(
                "name" to name,
                "email" to email,
                "phone" to phone,
                "patientId" to patientId,
                "service" to service,
                "date" to date,
                "time" to time,
                "message" to message,
            ),
            files = documents,
            fileFieldName = "files",
        )
    }

    suspend fun getPatient(patientId: String): JsonMap {
        return getJson("/patients/$patientId")
    }

    suspend fun updatePatientProfile(
        patientId: String,
        name: String,
        mobile: String,
        disease: String,
    ): JsonMap {
        return putJson(
            "/patients/$patientId",
            mapOf(
                "name" to name,
                "mobile" to mobile,
                "disease" to disease,
            ),
        )
    }

    suspend fun changePassword(oldPassword: String, newPassword: String, confirmPassword: String): JsonMap {
        return postJson(
            "/auth/change-password",
            mapOf(
                "oldPassword" to oldPassword,
                "newPassword" to newPassword,
                "confirmPassword" to confirmPassword,
            ),
        )
    }

    suspend fun uploadPatientProfileImage(patientId: String, image: UploadFile): JsonMap {
        return postMultipart(
            "/patients/$patientId/profile-image",
            fields = emptyMap(),
            files = listOf(image),
            fileFieldName = "image",
        )
    }

    suspend fun getServices(): List<JsonMap> = extractList(getJson("/services"))

    suspend fun getShopProducts(): List<JsonMap> = extractList(getJson("/shop/products"))

    suspend fun getPatientAppointmentRequests(patientId: String): List<JsonMap> {
        return extractList(getJson("/patients/$patientId/appointment-requests"))
    }

    suspend fun getMyShopOrders(): List<JsonMap> = extractList(getJson("/shop/orders/my"))

    suspend fun placeShopOrder(items: List<JsonMap>, note: String): JsonMap {
        return postJson(
            "/shop/orders",
            mapOf(
                "items" to items,
                "note" to note,
            ),
        )
    }

    suspend fun getPublicChatAgents(): List<JsonMap> = extractList(getJson("/public-chat-agents"))

    suspend fun getPublicChatConversation(conversationId: String): JsonMap {
        return getJson("/public-chat/conversations/$conversationId")
    }

    suspend fun startPublicChatConversation(
        agentId: String,
        visitorName: String,
        visitorContact: String,
        text: String,
        attachments: List<UploadFile>,
    ): JsonMap {
        return postMultipart(
            "/public-chat/conversations",
            fields = mapOf(
                "agentId" to agentId,
                "visitorName" to visitorName,
                "visitorContact" to visitorContact,
                "text" to text,
            ),
            files = attachments,
            fileFieldName = "attachments",
        )
    }

    suspend fun sendPublicChatMessage(
        conversationId: String,
        visitorName: String,
        text: String,
        attachments: List<UploadFile>,
    ): JsonMap {
        return postMultipart(
            "/public-chat/conversations/$conversationId/messages",
            fields = mapOf(
                "visitorName" to visitorName,
                "text" to text,
            ),
            files = attachments,
            fileFieldName = "attachments",
        )
    }

    fun resolveResourceUrl(pathOrUrl: String): String {
        val trimmed = pathOrUrl.trim()
        if (trimmed.isEmpty()) {
            return ""
        }
        if (trimmed.startsWith("http://") || trimmed.startsWith("https://")) {
            return trimmed
        }

        val baseUrl = resolveBaseUrl()
        return if (trimmed.startsWith("/")) {
            baseUrl + trimmed
        } else {
            "$baseUrl/$trimmed"
        }
    }

    private suspend fun getJson(path: String): JsonMap = requestJson("GET", path, null)

    private suspend fun postJson(path: String, payload: JsonMap): JsonMap = requestJson("POST", path, payload)

    private suspend fun putJson(path: String, payload: JsonMap): JsonMap = requestJson("PUT", path, payload)

    private suspend fun requestJson(method: String, path: String, payload: JsonMap?): JsonMap {
        return guardRequest {
            withContext(Dispatchers.IO) {
                val connection = openConnection(method, path)
                connection.setRequestProperty("Content-Type", "application/json")
                if (payload != null) {
                    connection.doOutput = true
                    DataOutputStream(connection.outputStream).use { output ->
                        output.write(JsonUtils.toJsonString(payload).toByteArray(StandardCharsets.UTF_8))
                    }
                }

                readJsonResponse(connection)
            }
        }
    }

    private suspend fun postMultipart(
        path: String,
        fields: Map<String, String>,
        files: List<UploadFile>,
        fileFieldName: String = "documents",
    ): JsonMap {
        return guardRequest {
            withContext(Dispatchers.IO) {
                val boundary = "----ommphysio${System.currentTimeMillis()}"
                val connection = openConnection("POST", path)
                connection.setRequestProperty("Content-Type", "multipart/form-data; boundary=$boundary")
                connection.doOutput = true

                DataOutputStream(connection.outputStream).use { output ->
                    fields.forEach { (name, value) ->
                        output.writeBytes("--$boundary\r\n")
                        output.writeBytes("Content-Disposition: form-data; name=\"${escapeMultipartValue(name)}\"\r\n\r\n")
                        output.write(value.toByteArray(StandardCharsets.UTF_8))
                        output.writeBytes("\r\n")
                    }

                    files.forEach { file ->
                        output.writeBytes("--$boundary\r\n")
                        output.writeBytes(
                            "Content-Disposition: form-data; name=\"${escapeMultipartValue(fileFieldName)}\"; " +
                                "filename=\"${escapeMultipartValue(file.name)}\"\r\n"
                        )
                        output.writeBytes("Content-Type: ${file.mimeType ?: guessMimeType(file.name)}\r\n\r\n")
                        output.write(file.bytes)
                        output.writeBytes("\r\n")
                    }

                    output.writeBytes("--$boundary--\r\n")
                    output.flush()
                }

                readJsonResponse(connection)
            }
        }
    }

    private suspend fun guardRequest(block: suspend () -> JsonMap): JsonMap {
        return try {
            block()
        } catch (error: ApiException) {
            throw error
        } catch (_: SocketTimeoutException) {
            throw ApiException("OPW is taking too long to respond. Please try again in a moment.")
        } catch (_: SSLException) {
            throw ApiException("Secure connection failed. Please try again shortly.")
        } catch (_: IOException) {
            throw ApiException("Unable to reach OPW right now. Please check your network and try again.")
        } catch (_: Exception) {
            throw ApiException("Unexpected response from OPW. Please try again.")
        }
    }

    private fun openConnection(method: String, path: String): HttpURLConnection {
        val connection = (URL(resolveBaseUrl() + path).openConnection() as HttpURLConnection).apply {
            requestMethod = method
            connectTimeout = REQUEST_TIMEOUT_MS
            readTimeout = REQUEST_TIMEOUT_MS
            doInput = true
            setRequestProperty("Accept", "application/json")
        }

        val token = storage.getPatientUser()?.string("token").orEmpty()
        if (token.isNotBlank()) {
            connection.setRequestProperty("Authorization", "Bearer $token")
        }

        if (connection is HttpsURLConnection) {
            connection.instanceFollowRedirects = true
        }

        return connection
    }

    private fun readJsonResponse(connection: HttpURLConnection): JsonMap {
        val statusCode = connection.responseCode
        val stream = if (statusCode in 200..299) connection.inputStream else connection.errorStream
        val body = stream?.use { input ->
            BufferedInputStream(input).readBytes().toString(StandardCharsets.UTF_8)
        }.orEmpty()

        val parsed = decodeBody(body)
        if (statusCode !in 200..299) {
            val message = parsed["message"]?.toString()
                ?: "Request failed with status $statusCode."
            throw ApiException(message, statusCode)
        }

        return parsed
    }

    private fun decodeBody(body: String): JsonMap {
        val trimmed = body.trim()
        if (trimmed.isEmpty()) {
            return emptyMap()
        }

        return when {
            trimmed.startsWith("{") -> JsonUtils.fromJsonObject(JSONObject(trimmed))
            trimmed.startsWith("[") -> mapOf("data" to JsonUtils.fromJsonArray(JSONArray(trimmed)))
            else -> throw ApiException("Unexpected response from OPW. Please try again.")
        }
    }

    private fun extractList(response: JsonMap): List<JsonMap> = response["data"].asJsonMapList()

    private fun resolveBaseUrl(): String {
        return BuildConfig.API_BASE_URL.trim().trimEnd('/')
    }

    private fun escapeMultipartValue(value: String): String = value.replace("\"", "\\\"")

    private fun guessMimeType(fileName: String): String {
        val lowerName = fileName.lowercase()
        return when {
            lowerName.endsWith(".pdf") -> "application/pdf"
            lowerName.endsWith(".png") -> "image/png"
            lowerName.endsWith(".jpg") || lowerName.endsWith(".jpeg") -> "image/jpeg"
            lowerName.endsWith(".webp") -> "image/webp"
            lowerName.endsWith(".doc") -> "application/msword"
            lowerName.endsWith(".docx") -> "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
            else -> "application/octet-stream"
        }
    }

    private companion object {
        const val REQUEST_TIMEOUT_MS = 25_000
    }
}
